packages/react/src/components/popover/useObserver.ts

```diff
@@ -0,0 +1,31 @@
+import { RefObject, useEffect, useRef } from 'react';
+
+/**
+ * Observes `targets` for intersection changes.
+ * @param targets `RefObject` of element to observe.
+ * @param onChange Called when the threshold changes. Either the element just became visible or the opposite.
+ * @param debounceTime Time in milliseconds to debounce events.
+ */
+export function useObserver<T extends Element>(
+  onChange: (entry: IntersectionObserverEntry) => void,
+  targets: RefObject<T>[],
+  debounceTime?: number
+) {
+  const timeout = useRef<number>();
+
+  useEffect(() => {
+    const elements = targets.map((t) => t.current).filter(Boolean) as T[];
+    if (!elements.length) return;
+
+    const observer = new IntersectionObserver(
+      ([entry]) => {
+        clearTimeout(timeout.current);
+        timeout.current = setTimeout(onChange, debounceTime, entry) as never;
```

Comment:
Not clear why we need a timeout here?

Reply:
To debounce if set.
For example if `debounceTime === 300` and multiple events trigger in under 300ms, `onChange` will only be called once.
We ended up not using `debounceTime` yet but I suspect I'll need to, in order to fix a bug with the "auto-optimal-positioning" feature coming next.